userops_ext.go

```diff
@@ -171,11 +183,41 @@ func no0xPrefix(value []byte) bool {
 // the Signature field. The Intent JSON is expected to be appended to the
 // signature value within the Signature field. The signature has a fixed length
 // of 132 characters with '0x' prefix.
+// It also takes into account cross-chain calldata specific format which is
+// prioritized and is in the following format:
+//
+// [2 bytes opType (0xFFFF)]
+// [2 bytes length of intent JSON]
+// [Intent JSON]
+// [1 byte hash list length (N)]
+// [Hash List Entry]
 //
 // Returns:
 //   - string: The extracted JSON string.
 //   - bool: A boolean indicating if a valid JSON was found.
 func (op *UserOperation) extractIntentJSON() (string, bool) {
+
+	// Try to check if it is a cross chain calldata
+	// If it is, extract the intent json from there first
+	dataBytes := []byte(op.CallData)
+
+	if len(dataBytes) >= 4 &&
+		binary.BigEndian.Uint16(dataBytes[:OpTypeLength]) == CrossChainMarker {
+
+		intentLengthBytes := dataBytes[2:4]
+		intentLength := int(binary.BigEndian.Uint16(intentLengthBytes))
+
+		if len(dataBytes) >= 4+intentLength {
+			intentJSON := string(dataBytes[4 : 4+intentLength])
+
+			if len(dataBytes) > 4+intentLength {
+				return intentJSON, true
+			}
+		}
+
+		return "", false
+	}
+
 	// Try to extract Intent JSON from CallData field
 	if intentJSON, ok := ExtractJSONFromField(string(op.CallData)); ok {
```

Comment:
Enhance `ExtractJSONFromField` to support cross-chain ops
- `[2 bytes opType (0xFFFF)]`
- `[2 bytes length of intent JSON]`
- `[Intent JSON]`
- `[1 byte hash list length (N)]`
- `[`**Hash List Entry**`]`